File: Core/src/main/java/com/inverce/mod/core/threadpool/UIScheduler.java
package com.inverce.mod.core.threadpool;

import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;

import java.util.concurrent.Callable;
import java.util.concurrent.Executor;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.TimeUnit;


public class UIScheduler implements Executor {
    protected static ScheduledThreadPoolExecutor scheduler;
    @NonNull
    protected final Handler uiHandler;

    public UIScheduler() {
        uiHandler = new Handler(Looper.getMainLooper());
        if (scheduler == null) {
            scheduler = new ScheduledThreadPoolExecutor(1, new NamedThreadPool("UI-Scheduler"));
            scheduler.setKeepAliveTime(2, TimeUnit.SECONDS);
            scheduler.allowCoreThreadTimeOut(true);
        }
    }

    @NonNull
    public ScheduledFuture<?> schedule(@NonNull Runnable command, long delay, @NonNull TimeUnit unit) {
        return scheduler.schedule(wrap(command), delay,  unit);
    }

    @NonNull
    public <V> ScheduledFuture<V> schedule(@NonNull Callable<V> callable, long delay, @NonNull TimeUnit unit) {
        return scheduler.schedule(callable, delay, unit);
    }

    @NonNull
    public ScheduledFuture<?> scheduleAtFixedRate(@NonNull Runnable command, long initialDelay, long period, @NonNull TimeUnit unit) {
        return scheduler.scheduleAtFixedRate(wrap(command), initialDelay, period, unit);
    }

    @NonNull
    public ScheduledFuture<?> scheduleWithFixedDelay(@NonNull Runnable command, long initialDelay, long delay, @NonNull TimeUnit unit) {
        return scheduler.scheduleWithFixedDelay(wrap(command), initialDelay, delay, unit);
    }

    private Runnable wrap(@NonNull Runnable runnable) {
        //noinspection Convert2Lambda
        return new Runnable() {
            @Override
            public void run() {
                execute(runnable);
            }
        };
    }

    @Override
    public void execute(@NonNull Runnable command) {
        uiHandler.post(command);
    }
}

File: Core/src/main/java/com/inverce/mod/v2/core/threadpool/UIScheduler.kt
package com.inverce.mod.v2.core.threadpool

import android.os.Handler
import android.os.Looper
import java.util.concurrent.*

open class UIScheduler : Executor {
    protected companion object {
        @set:Synchronized
        @get:Synchronized
        private var count = 0
    }

    protected val uiHandler: Handler = Handler(Looper.getMainLooper())
    protected val scheduler: ScheduledThreadPoolExecutor by lazy {
        ScheduledThreadPoolExecutor(1, NamedThreadPool("UI-Scheduler-${count++}")).apply {
            setKeepAliveTime(2, TimeUnit.SECONDS)
            allowCoreThreadTimeOut(true)
        }
    }

    fun schedule(command: Runnable, delay: Long, unit: TimeUnit): ScheduledFuture<*> {
        return scheduler.schedule(wrap(command), delay, unit)
    }

    fun <V> schedule(callable: Callable<V>, delay: Long, unit: TimeUnit): ScheduledFuture<V> {
        return scheduler.schedule(callable, delay, unit)
    }

    fun scheduleAtFixedRate(command: Runnable, initialDelay: Long, period: Long, unit: TimeUnit): ScheduledFuture<*> {
        return scheduler.scheduleAtFixedRate(wrap(command), initialDelay, period, unit)
    }

    fun scheduleWithFixedDelay(command: Runnable, initialDelay: Long, delay: Long, unit: TimeUnit): ScheduledFuture<*> {
        return scheduler.scheduleWithFixedDelay(wrap(command), initialDelay, delay, unit)
    }

    private fun wrap(runnable: Runnable): Runnable {
        return Runnable { execute(runnable) }
    }

    override fun execute(command: Runnable) {
        uiHandler.post(command)
    }
}